### PR TITLE
experimental: creds2

### PR DESCRIPTION
**Pull request by pkoch**
This is just a sketch of the implementation to make sure all the wires align.

See also:
- https://github.com/idos-network/idos-sdk-js/pull/340
- https://github.com/idos-network/idos-schema/pull/55
- https://github.com/idos-network/idos-kgw/pull/25

### CODE REVIEW

File: schema.production.kf [hunk 1]
```diff
@@ -461,6 +622,70 @@ action share_credential_through_dag (
...
+        SELECT 1 from credentials2
+            WHERE id = $original_credential_id
+            AND human_id = $human_id
+    ) THEN
+        error('The original credential does not belong to the human')
+    END;
+
+    $owner_verified = idos.verify_owner($dag_owner, $dag_grantee, $id, $dag_locked_until, $dag_signature);
+    SELECT CASE WHEN $owner_verified != 1 THEN
+        error('the signature is invalid')
```
Comment (sapience) [inline schema.production.kf:651]: As now we have a loooot of signatures i would name the error as `the dag_signature is invalid`

File: schema.production.kf [hunk 2]
```diff
@@ -509,6 +734,66 @@ procedure share_credential_by_write_grant(
...
+    SELECT CASE WHEN get_write_grant_id($human_id)::text != get_credential2_inserter($original_credential_id) THEN
+        error('you can share only original credentials you created')
+    END;
+
+    SELECT CASE WHEN (SELECT count(1) FROM internal_access_grants WHERE id = uuid_generate_v5('31276fd4-105f-4ff7-9f64-644942c14b79'::uuid, format('%s-%s-%s', $grantee::text, $original_credential_id::text, $locked_until::text))) > 0 THEN
+        error('a grant with the same grantee, original_credential_id, and locked_until already exists')
+    END;
+
+    SELECT CASE WHEN $public_notes != '' THEN
+        error('shared credentials cannot have public_notes, it must be an empty string')
```
Comment (sapience) [inline schema.production.kf:769]: Oh. From this perspective proposed structure of `credentials2` looks not very good. If we eventually will keep only new credentials, then maybe it is worth to think about split original and shared credentials to different tables?

File: schema.production.kf [hunk 3]
```diff
@@ -42,6 +42,20 @@ table credentials {
...
 }
 
+table credentials2 {
+    id uuid primary,
+    human_id uuid notnull,
+    public_notes text notnull,
+    public_notes_signature text notnull,
+    content text notnull,
+    broader_signature text notnull,
+    encryption_public_key text notnull, // public key of encryptor
```
[the remark this answers is not in the document]
Comment (pkoch) [reply schema.production.kf:50]: It's either the issuer, or the human with a throw-away signing key.

File: schema.production.kf [hunk 4]
```diff
@@ -42,6 +42,20 @@ table credentials {
...
+    issuer text notnull,
```
[the remark this answers is not in the document]
Comment (pkoch) [reply schema.production.kf:51]: `issuer_authentication_public_key`, at least, to disambiguate the key usage.

File: schema.production.kf [hunk 5]
```diff
@@ -385,6 +486,36 @@ action edit_credential($id, $credential_type, $credential_level, $credential_sta
...
+, $issuer
+) public {
+    SELECT CASE
+        WHEN EXISTS (
+            SELECT 1 from credentials2 AS c
+                INNER JOIN shared_credentials AS sc on c.id = sc.duplicate_id
+                WHERE c.id = $id
+                AND c.human_id=(SELECT DISTINCT human_id FROM wallets WHERE (wallet_type = 'EVM' AND address=@caller COLLATE NOCASE)
+                    OR (wallet_type = 'NEAR' AND public_key = @caller))
+        ) THEN ERROR('Can not edit shared credential') END;
```
[the remark this answers is not in the document]
Comment (pkoch) [reply schema.production.kf:505]: We went in a different direction: only the original credential has public notes, and the shared ones need to join with the original row to get them.

File: schema.production.kf (hunk 1 above)
Comment (pkoch) [reply schema.production.kf:653]: Great point. :ok_hand:

File: schema.production.kf (hunk 2 above)
Comment (pkoch) [reply schema.production.kf:769]: Yeah, that also went through my head. The only reason I didn't do it is that we can always do that change after the fact. For now, I'm foreseeing enough fun on the migration with just the schema changes we already have.

File: schema.production.kf (hunk 3 above)
[the remark this answers is not in the document]
Comment (sapience) [reply schema.production.kf:50]: So `issuer` and `encryption_public_key` can be different?

[the remark this answers is not in the document]
Comment (pkoch) [reply schema.production.kf:50]: They have to be different: one is a signing key, the other is an encryption key.

Maybe I'm not understanding your question right? :thinking:

[the remark this answers is not in the document]
Comment (sapience) [reply schema.production.kf:50]: Sorry, my bad. The question was dumb.

**Pull request closed without merging**